### PR TITLE
Drop collections between tests, fix a couple of issues with alternative drivers

### PR DESCRIPTION
**Pull request by vkarpov15**
**Summary**

Re: #11999, we get some issues with dropping indexes and deleting documents at the same time in tests. The reason why we do `dropIndexes()` and `deleteMany()` in between test cases is that MongoDB's WiredTiger storage engine used to be incredibly slow when dropping databases and collections - it used to take over 60 seconds to `dropDatabase()`. These days the difference isn't anywhere near that bad. On my local, it takes about 4 seconds to run document tests with WT using `dropIndex()` and `deleteMany()`, and 11 seconds with `collection.drop()`. We also typically use the in-memory storage engine for tests anyway.

Also fixed a couple of issues that we ran into: `bulkWrite()` and other collection methods can throw synchronous errors, so we need to be careful to catch those.

**Examples**

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -3575,13 +3575,17 @@ Model.bulkWrite = function(ops, options, callback) {
...
+          if (error) {
+            return cb(error);
+          }
 
-        cb(null, res);
-      });
+          cb(null, res);
+        });
+      } catch (err) {
+        return cb(err);
```
Comment (AbdelrahmanHafez) [inline lib/model.js:3587]: I've always didn't like the fact that bulkWrite could throw both sync and async errors, but I accepted it as a fact.
Glad this is changing now.

**Review by AbdelrahmanHafez — CHANGES_REQUESTED** [review]
We have a few failing tests, otherwise LGTM. Thanks. 👍

**Review by IslandRhythms — APPROVED** [review]

File: test/model.test.js [hunk 2]
```diff
@@ -4229,7 +4247,12 @@ describe('Model', function() {
       });
 
 
-      await Person.collection.drop();
+      await Person.collection.drop().catch(err => {
+        if (err.codeName === 'NamespaceNotFound') {
```
Comment (AbdelrahmanHafez) [inline test/model.test.js:4251]: Should mongoose, generally, ignore this error and return a fake result?
If I drop a collection and that connection does not exist, I wouldn't mind if I just receive a response saying that it's dropped.

What do you think? @vkarpov15

Comment (vkarpov15) [reply test/model.test.js:4251]: @AbdelrahmanHafez we may consider that for a future release, but it's a pretty low priority.

Comment (AbdelrahmanHafez) [reply test/model.test.js:4251]: I agree it's a low priority.

**Pull request merged**